lib/twilioconnection.js

```diff
@@ -12,23 +12,40 @@ let nInstances = 0;
   TwilioConnection states
   -----------------------
 
-  +--------------+       +----------+
-  |  connecting  | ----> |  closed  |
-  +--------------+       +----------+
-         |                    ^
-         v                    |
-     +--------+               |
-     |  open  | ---------------
-     +--------+
+       ------------------------------------------
+       |                                        |
+       |                                        v
+  +---------+       +--------------+       +----------+
+  |  early  | ----> |  connecting  | ----> |  closed  |
+  +---------+       +--------------+       +----------+
+    ^                     | ^ |                 ^ ^
+    | --------------------- | |                 | |
+    | | --------------------- |                 | |
+    | | | --------------------|------------------ |
+    | v | |                   v                   |
+  +----------+           +--------+               |
+  |   wait   |           |  open  | ---------------
+  +----------+           +--------+
  */
 
 const states = {
   closed: [],
-  connecting: ['closed', 'open'],
-  open: ['closed']
+  connecting: ['closed', 'open', 'wait'],
+  early: ['closed', 'connecting'],
+  open: ['closed'],
+  wait: ['closed', 'connecting', 'early']
 };
 
+const events = {
+  closed: 'close',
+  open: 'open',
+  wait: 'wait'
+};
+
+const TCMP_VERSION = 2;
+
 const DEFAULT_MAX_CONSECUTIVE_MISSED_HEARTBEATS = 3;
+const DEFAULT_MAX_CONSECUTIVE_FAILED_HELLOS = 3;
```

Comment:
Nitpick: I'm not sure that this value needs to be configurable. The idea is that hellos are just heartbeats used to establish the initial connection so you could reuse `DEFAULT_MAX_CONSECUTIVE_MISSED_HEARTBEATS`.

Reply:
@ceaglest You're right. We don't override the default behavior when using TwilioConnection in the SDK. However, we use this for unit tests.